recipe-client-addon/lib/RecipeRunner.jsm

```diff
@@ -39,88 +39,147 @@ Cu.importGlobalProperties(["fetch"]);
 this.EXPORTED_SYMBOLS = ["RecipeRunner"];
 
 const log = LogManager.getLogger("recipe-runner");
-const prefs = Services.prefs.getBranch("extensions.shield-recipe-client.");
 const TIMER_NAME = "recipe-client-addon-run";
-const RUN_INTERVAL_PREF = "run_interval_seconds";
-const FIRST_RUN_PREF = "first_run";
 const PREF_CHANGED_TOPIC = "nsPref:changed";
 
+const TELEMETRY_ENABLED_PREF = "datareporting.healthreport.uploadEnabled";
+
+const SHIELD_PREF_PREFIX = "extensions.shield-recipe-client";
+const RUN_INTERVAL_PREF = `${SHIELD_PREF_PREFIX}.run_interval_seconds`;
+const FIRST_RUN_PREF = `${SHIELD_PREF_PREFIX}.first_run`;
+const SHIELD_ENABLED_PREF = `${SHIELD_PREF_PREFIX}.enabled`;
+const DEV_MODE_PREF = `${SHIELD_PREF_PREFIX}.dev_mode`;
+const API_URL_PREF = `${SHIELD_PREF_PREFIX}.api_url`;
+const LAZY_CLASSIFY_PREF = `${SHIELD_PREF_PREFIX}.experiments.lazy_classify`;
+
+const PREFS_TO_WATCH = [
+  RUN_INTERVAL_PREF,
+  TELEMETRY_ENABLED_PREF,
+  SHIELD_ENABLED_PREF,
+  API_URL_PREF,
+];
+
 this.RecipeRunner = {
   async init() {
-    if (!this.checkPrefs()) {
-      return;
-    }
+    this.enabled = null;
+    this.checkPrefs();  // sets this.enabled
+    this.watchPrefs();
+
+    // Run if enabled immediately on first run, or if dev mode is enabled.
+    const firstRun = Services.prefs.getBoolPref(FIRST_RUN_PREF);
+    const devMode = Services.prefs.getBoolPref(DEV_MODE_PREF);
 
-    // Run immediately on first run, or if dev mode is enabled.
-    if (prefs.getBoolPref(FIRST_RUN_PREF) || prefs.getBoolPref("dev_mode")) {
+    if (this.enabled && (devMode || firstRun)) {
       await this.run();
-      prefs.setBoolPref(FIRST_RUN_PREF, false);
     }
+    if (firstRun) {
+      Services.prefs.setBoolPref(FIRST_RUN_PREF, false);
+    }
+  },
 
+  enable() {
+    if (this.enabled) {
+      return;
+    }
     this.registerTimer();
+    this.enabled = true;
   },
 
-  registerTimer() {
-    this.updateRunInterval();
-    CleanupManager.addCleanupHandler(() => timerManager.unregisterTimer(TIMER_NAME));
+  disable() {
+    if (!this.enabled) {
+      return;
+    }
+    this.unregisterTimer();
+    this.enabled = false;
+  },
+
+  /** Watch for prefs to change, and call this.observer when they do */
+  watchPrefs() {
+    for (const pref of PREFS_TO_WATCH) {
+      Services.prefs.addObserver(pref, this);
+    }
 
-    // Watch for the run interval to change, and re-register the timer with the new value
-    prefs.addObserver(RUN_INTERVAL_PREF, this);
-    CleanupManager.addCleanupHandler(() => prefs.removeObserver(RUN_INTERVAL_PREF, this));
+    CleanupManager.addCleanupHandler(this.unwatchPrefs.bind(this));
```

Comment:
Hm. Are you sure this works? I thought this would give you a different reference, when you need the same for both add/remove... I've always done this like:

```
this.unwatchPrefs = this.unwatchPrefs.bind(this);
addListener(this.unwatchPrefs);
/* later... */
removeListener(this.unwatchPrefs);
```

Reply:
You're right that if we needed to remove a listener later, this would be a problem. But we never remove cleanup handlers, so we won't need this reference again. CleanupHandler is an Shield-ism. this will be the same in the callback (unwatchPrefs), soServices.prefs.removeObserver(pref, this)` (line 107) will work fine.